inc/saitam.h

```diff
@@ -34,866 +34,1619 @@
+{
     /**
-     * @brief Statistics reporting mode.
-     *
-     * When not specified, reports in number of bytes (DEFAULT)
-     *
-     * @type sai_tam_reporting_mode_t
-     * @flags CREATE_AND_SET
-     * @default SAI_TAM_REPORTING_MODE_BYTES
+     * @brief No Authentication
      */
-    SAI_TAM_ATTR_BUFFER_REPORTING_MODE,
+    SAI_TAM_TRANSPORT_AUTH_TYPE_NONE,
 
     /**
-     * @brief Buffer Tracker Mode
-     *
-     * Specifies whether the Chip should track the peak values of the
-     * buffers or current usage values (DEFAULT).
-     *
-     * @type sai_tam_tracking_mode_t
-     * @flags CREATE_AND_SET
-     * @default SAI_TAM_TRACKING_MODE_CURRENT
+     * @brief Authenticate using SSL
      */
-    SAI_TAM_ATTR_BUFFER_TRACKING_MODE,
+    SAI_TAM_TRANSPORT_AUTH_TYPE_SSL,
 
     /**
-     * @brief Buffers/Statistics for tracking using this object
-     *
-     * Specifies the Statistics/Types for tracking. If not specified, all
-     * supported buffers are included for tracking (DEFAULT).
-     *
-     * A statistic can't be tracked by more than one TAM Objects.
-     *
-     * @type sai_object_list_t
-     * @flags CREATE_AND_SET
-     * @objects SAI_OBJECT_TYPE_TAM_STAT
-     * @default empty
+     * @brief Authenticate using TLS
      */
-    SAI_TAM_ATTR_TRACKING_OPTIONS,
+    SAI_TAM_TRANSPORT_AUTH_TYPE_TLS
+
+} sai_tam_transport_auth_type_t;
+
+/**
+ * @brief Transport object Attributes
+ */
+typedef enum _sai_tam_transport_attr_t
+{
 
     /**
-     * @brief Default Transporter
-     *
-     * Provides a default snapshot transporter object for the Tracker.
-     * When a snapshot is made, this transporter will be used
-     * to 'copy' the data to the 'transporter-desired' location.
-     *
-     * In the absence of a transporter, the tracker will copy the
-     * data to the local CPU transporter.
-     *
-     * @type sai_object_id_t
-     * @flags CREATE_AND_SET
-     * @objects SAI_OBJECT_TYPE_TAM_TRANSPORTER
-     * @allownull true
-     * @default SAI_NULL_OBJECT_ID
+     * @brief Start of Attributes
      */
-    SAI_TAM_ATTR_TRANSPORTER,
+    SAI_TAM_TRANSPORT_ATTR_START,
 
     /**
-     * @brief Clear all Thresholds
+     * @brief Transport type
      *
-     * If this attribute is specified and set to true, then
-     * the following actions take place for each of the created threshold objects
-     * 1. The values specified via the SAI_TAM_THRESHOLD_ATTR_LEVEL attribute are
-     *    removed
-     * 2. Threshold monitoring is disabled
-     *
-     * @type bool
-     * @flags CREATE_AND_SET
-     * @default false
+     * @type sai_tam_transport_type_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
      */
-    SAI_TAM_ATTR_CLEAR_ALL_THRESHOLDS,
+    SAI_TAM_TRANSPORT_ATTR_TRANSPORT_TYPE = SAI_TAM_TRANSPORT_ATTR_START,
 
     /**
-     * @brief Total Number of counters supported
+     * @brief Transport src port
+     * Value of -1 can be used a hint to compute ephemeral value
```

Comment:
Looking up the dictionary definition of "ephemeral", not sure that is a strong enough hint to what this does.

How about:
Value of -1 (all 1s) indicates that any value may be transmitted. The transmitted value may vary in different reports sent on the same transport, for example to introduce entropy in equal cost path selection.

Reply:
Quoting from Jai's reply:

> [Jai] You are looking for functionality similar to VxLAN source UDP port and the property defined for entropy is by using ephemeral port which can be either hash on multiple fields or a static value.

> Please refer to IETF draft.

> RFC 7348 - Virtual eXtensible Local Area Network (VXLAN):

We agree on the technical behavior. Where we are disagreeing is with regard to terminology. Quoting from your reference of RFC 7348:

> Source Port:  It is recommended that the UDP source port number be calculated using a hash of fields from the inner packet -- one example being a hash of the inner Ethernet frame's headers. This is to enable a level of entropy for the ECMP/load-balancing of the VM-to-VM traffic across the VXLAN overlay. When calculating the UDP source port number in this manner, it is RECOMMENDED that the value be in the dynamic/private port range 49152-65535 [RFC6335].

There is no mention of the word "ephemeral" in this text. I conclude that "ephemeral" is your terminology, not industry standard terminology. Meanwhile the RFC 7348 text does mention "entropy", "ECMP", "load-balancing", all of which are industry standard terminology. We should be describing this using the industry standard terminology.

Reply:
Quoting from Jai's latest reply:

> Please look at this RFC and it mentions vxlan src UDP port as ephemeral port in the entire draft.

> https://tools.ietf.org/id/draft-saum-nvo3-pmtud-over-vxlan-04.html

You are referring to a relatively recent document still in draft status, that does not deal directly with the functionality we are referring to.

When I read the term "ephemeral" I had to go look it up. From the dictionary definition it was not at all obvious what behavior it was referring to. I stand by my recommendation to use industry standard terminology from your suggested reference, RFC 7348, i.e. "entropy" and "ECMP" or "equal cost". Everyone understands what the latter terms mean.